### PR TITLE
Quarkus 3 jakarta roadmap#16

### PR DESCRIPTION
**Pull request by gitgabrio**
Fixes https://github.com/IBM/kie-roadmap/issues/16

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] mandrel</b>

 - for <b>mandrel lts checks</b>  
  Run native checks against Mandrel image and quarkus lts branch
  Please add comment: <b>Jenkins run mandrel-lts</b>

- for a <b>specific mandrel lts check</b>  
  Run native checks against Mandrel image and quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] mandrel-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: jpmml-migration-recipe/src/main/java/org/kie/openrewrite/recipe/jpmml/JPMMLCodeRecipe.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    private static Properties CHANGED_INSTANTIATIONS;
+
+    static {
+        try (InputStream input = JPMMLRecipe.class.getResourceAsStream ("/changed_instantiation.properties")) {
+            CHANGED_INSTANTIATIONS = new Properties();
+            // load a properties file
+            CHANGED_INSTANTIATIONS.load(input);
+        } catch (IOException ex) {
+            ex.printStackTrace();
```
Comment (pibizza) [inline jpmml-migration-recipe/src/main/java/org/kie/openrewrite/recipe/jpmml/JPMMLCodeRecipe.java]: I think you should use a logger here.

File: pom.xml [hunk 2]
```diff
@@ -200,6 +200,7 @@
         <module>kie-maven-plugin</module>
         <module>kie-archetypes</module>
         <module>drools-drl-quarkus-extension</module>
+        <module>jpmml-migration-recipe</module>
```
Comment (pibizza) [inline pom.xml:203]: We probably want - eventually - to contribute this  recipe back to the openrewrite project. At that point we will have to split away from this package.

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                "        return \"Hello from com.yourorg.FooBar!\";\n" +
+                "    }\n" +
+                "}";
+        String methodTested = "System.out.println";
+        J.MethodInvocation toTest = getMethodInvocationFromClassSource(classTested, methodTested).iterator().next();
+        assertThat(toTest).isNotNull();
+        ExecutionContext executionContext = getExecutionContext(null);
+        J.MethodInvocation retrieved = jpmmlVisitor.visitMethodInvocation(toTest, executionContext);
+        String expected = "System.out.println(\"OUTPUT_\")";
+        assertThat(retrieved).isNotNull()
```
Comment (pibizza) [inline jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: Use the fluent formatting in a coherent way. IsNotNull should be on a newline.

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                "    }\n" +
+                "\n" +
+                "}";
+        String classInstantiated = "org.dmg.pmml.ScoreDistribution";
+        J.NewClass toTest = getNewClassFromClassSource(classTested, classInstantiated).iterator().next();
+        assertThat(toTest).isNotNull();
+        ExecutionContext executionContext = getExecutionContext(null);
+        J.NewClass retrieved = jpmmlVisitor.visitNewClass(toTest, executionContext);
+        String expected = "new ComplexScoreDistribution()";
+        assertThat(retrieved).isNotNull()
```
Comment (pibizza) [inline jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: Use fluent formatting in a coherent way. IsNotNull should be a on a separate line.

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import static org.kie.openrewrite.recipe.jpmml.JPMMLVisitor.TO_MIGRATE_MESSAGE;
+
+class JPMMLVisitorTest {
+
+    private JPMMLVisitor jpmmlVisitor;
+    private static Properties CHANGED_INSTANTIATIONS;
+
+    @BeforeAll
+    public static void setup() throws IOException {
+        try (InputStream input = JPMMLRecipe.class.getResourceAsStream ("/changed_instantiation.properties")) {
```
Comment (pibizza) [inline jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: Why loading this resource from a property file? I think this configuration information is supposed to be in rewrite.yml file.

**Review by pibizza — COMMENTED** [review]
Some comments:
1) probably we want to make it a separate recipe to submit to the openrewrite project
2) I think we should not use property files for configuration, if I understand correctly this should go into rewrite.yml file
3) we should use a logger instead of printStackTrace.

File: pom.xml (hunk 2 above)
Comment (gitgabrio) [reply pom.xml:203]: @pibizza 👍

File: jpmml-migration-recipe/src/main/java/org/kie/openrewrite/recipe/jpmml/JPMMLCodeRecipe.java (hunk 1 above)
Comment (gitgabrio) [reply jpmml-migration-recipe/src/main/java/org/kie/openrewrite/recipe/jpmml/JPMMLCodeRecipe.java]: done

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java (hunk 3 above)
Comment (gitgabrio) [reply jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: done

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java (hunk 4 above)
Comment (gitgabrio) [reply jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: done

File: jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java (hunk 5 above)
Comment (gitgabrio) [reply jpmml-migration-recipe/src/test/java/org/kie/openrewrite/recipe/jpmml/JPMMLVisitorTest.java]: @pibizza 
Done!

**Comment by gitgabrio** [comment]
@pibizza @radtriste 
I've simplified implementation.

1. the base recipe is declarative, named `org.kie.openrewrite.recipe.jpmml.JPMMLRecipe`
2. the "rewrite.yml" include configuration of ChangeType and of JPMMLCodeRecipe
3. JPMMLCodeRecipe is responsible to make specifically needed modification, via the JPMMLVisitor
4. if other custom-tailored modifications have to be made, they should be done inside JPMMLVisitor

To test the full roundtrip, I created an external project with the following (pom) configuration

`<plugin>
        <groupId>org.openrewrite.maven</groupId>
        <artifactId>rewrite-maven-plugin</artifactId>
        <version>4.38.0</version>
        <configuration>
          <activeRecipes>
            <activeRecipe>org.kie.openrewrite.recipe.jpmml.JPMMLRecipe</activeRecipe>
          </activeRecipes>
        </configuration>
        <dependencies>
          <dependency>
            <groupId>org.kie</groupId>
            <artifactId>jpmml-migration-recipe</artifactId>
            <version>8.33.0-SNAPSHOT</version>
          </dependency>
        </dependencies>
      </plugin>`
      
      The only missing bit is that after the conversion, the resultant code is broken, since it would need the new jpmml version. So, I think that also the change of the jpmml model version should be done inside this recipe

**Review by pibizza — APPROVED** [review]
Fine for me. Just remember to change the readme file - it still mentions the old  properties files that have been removed.

**Comment by gitgabrio** [comment]
> Fine for me. Just remember to change the readme file - it still mentions the old properties files that have been removed.

Done. Also add/fixed minor details

**Pull request merged**